content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,242 @@
+
+Now a new requirement is coming, we need to add a new method called `cover`, and it also has author first name and last name inside. We can follow the pattern we just made. Let's compare before and after the refactoring of the book class.
+#### Old pattern without refactoring
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+end
+ 
+def cover(first_name, last_name)
+  "#{title}\n\n\n#{last_name}, #{first_name}"
+end
+```
+#### After refactor
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+ 
+def cover(author)
+  "#{title}\n\n\n#{author.full_name}"
+end
+```
+The cover method generates a string that has the book title and author's full name inside. After the refactoring, we can reuse the author's full name method again. Once we build a good pattern, the new code can follow it easily and keep the project clean. But if you didn't pay attention, the code getting rust rapidly and becomes the developer's nightmare. 
+
+You may also notice that both `summary` and `cover` need the same parameter, author, it's the smell of refactoring. The author information should be one of the attributes of the book object, instead of passing as a parameter when we call the book method. So we can refactor the book class again.
+
+```rb
+class Book
+  attr_reader :title, :publish_year, :author
+  
+  def initialize(title, publish_year, author)
+    @title = title
+    @publish_year = publish_year
+    @author = author
+  end
+ 
+  def summary
+    "#{title} - #{author.full_name} (#{publish_year})"
+  end
+  
+  def cover
+    "#{title}\n\n\n#{author.full_name}"
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+book = Book.new("Ruby is the best", "2020", bater)
+book.summary
+=> "Ruby is the best - Chen, Bater (2020)"
+book.cover
+=> "Ruby is the best\n\n\nChen, Bater"
+```
+Both `summary` and `cover` method's output remain the same, but now we can call it freely without any parameter once the book object is prepared. In ruby, we can skip parentheses if there is no parameter.
+
+This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.
+
+## Open-closed principle
+Reducing the parameter from two to zero is not the only benefit we got from this refactoring, this new design is open for extension but closed for modification. It follows the [open-closed principle][3].
+
+Now we have a new requirement coming again, some authors have the middle name. In the new version, it would be very easy to adapt to the change, we don't need to touch the book class anymore. The new author class would refactor like this:
```

Comment:
```suggestion
Let's say we have a new requirement again, this time, some authors now have a middle name. In this new version, it would be very easy to adapt our code because we don't need to touch the book class anymore. The new author class would refactored like this:
```